### ISSUE TITLE
Add Flysystem V2 Support

### ISSUE THREAD

**Issue by dieter-pollier** [sulu/sulu#5701]
| Q | A
| --- | ---
| Sulu Version | 2.2
| Browser Version | Browser name and version

# Actual Behavior

When I configure S3 as external media storage, I get the following error on the cli:
`Unrecognized option "s3" under "sulu_media.storages". Available option is "local".`

# Expected Behavior

Sulu should recognize the configuration

# Steps to Reproduce

I followed the docs: https://docs.sulu.io/en/2.2/cookbook/external-media-storage.html
* Installed the packages flysystem and flysystem-aws-s3-v3
* Added the configuration yaml-snippet

# Possible Solutions

If you have already ideas how to solve the issue, add them here.
(remove this section if not needed)

**Comment by nnatter** [comment]
Hey, thanks for reporting. The `s3` option is only available if the `League\Flysystem\AwsS3v3\AwsS3Adapter` is found:
https://github.com/sulu/sulu/blob/0a30e7978638c8c0192768a468f324f34c0ed0ea/src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php#L210-L229

Does this class exist in your project?

**Comment by dieter-pollier** [comment]
Thanks for your reply. It seems that the name of the class has been changed in version 2.x: https://github.com/thephpleague/flysystem-aws-s3-v3/tree/2.x.
I installed using the command: `composer require league/flysystem:^1.0.0 league/flysystem-aws-s3-v3:^1.0.0`, and now the config gets loaded.

**Comment by nnatter** [comment]
Thanks for the information! I will create a PR for the documentation 🙂

**Comment by alexander-schranz** [comment]
I will keep this issue open as I think we should find a way in future to support the new [flysystem version](https://flysystem.thephpleague.com/v2/docs/advanced/upgrade-to-2.0.0/). We currently only support v1. We maybe need at current state add a `conflict` for version 2 package.

**Comment by nnatter** [comment]
Reminder: Revert https://github.com/sulu/sulu-docs/pull/604 when this is solved 🙂

**Comment by ju1ius** [comment]
Hi, any news on this (or possible workaround)?
`league/flysystem` and `league/flysystem-aws-s3-v3` are now at version `^3.0`... 🤔

**Comment by RafaelKr** [comment]
@alexander-schranz we just had a quick convo about jackalope versioning on another issue. You told me v3 is on the horizon. I don't want to spam that issue with off-topic, but maybe I have something interesting to look into for v3.0 which is somehow related to this issue.

There's an extremely good library for file abstraction: https://rekalogika.dev/file/intro and https://rekalogika.dev/file-bundle/intro

"High-level file abstraction library built on top of Flysystem. It lets you work with file objects in an object-oriented manner. A file object represents a file in a Flysystem filesystem. It can be a local file or a file in a cloud storage, the library lets you work with them in the same way."

They provide:
```
rekalogika/file: The core library. It provides the file abstraction and metadata support.
rekalogika/file-bundle: Integrates the library with Symfony.
rekalogika/file-association: Provides support for associating files with
rekalogika/file-association-entity: Utilities for handling files inside domain entities.
rekalogika/file-contracts: Contains the interfaces and contracts used by the library.
rekalogika/file-derivation: Library for creating derived files.
rekalogika/file-filepond: Form for uploading files utilizing the FilePond Javascript library.
rekalogika/file-image: Provides image resizing filter.
rekalogika/file-metadata: Classes that represent file metadata.
rekalogika/file-metadata-contracts: Contains additional interfaces
rekalogika/file-null: Implementation of the null object pattern.
rekalogika/file-oneup-uploader-bridge: Adapter for OneupUploaderBundle.
rekalogika/file-server: Temporary URL server for files.
rekalogika/file-symfony-bridge: Adapter for Symfony HttpFoundation, Form, and
rekalogika/file-zip: Streams files as a ZIP file to the client.
```

As I said in the other issue, I just started working with Sulu and have not worked with media/files yet. But working with the rekalogika file implementation was a breeze! Being able to associate Files with database entities and directly doing transformations/derivations (filters, resize, scale, ...) directly from the entity was such a game changer for me.
Also using rekalogika/file-filepond to directly integrate [FilePond](https://pqina.nl/filepond/), an awesome library for file uploads, was very easy.

So maybe it's worth checking out the library or at least some of their concepts. If it's not interesting to you, sorry for bothering.

**Comment by alexander-schranz** [comment]
We currently will not add more dependencies to the Sulu core specially as I don't see a real advantages here for our usecase. Also specially why they may be maintainance of your linked repos is currently good it does not mean that it is supported well in 2-4 years where we still support version of Sulu 3.0 and the past showed as with packages like goodby csv, ongr elasticsearch, ..., which required us to fork things and take over the maintainance of them.

We are definitly happy with Flysystem and Imagine at current state. With flysystem we definitly know that it will stay well maintained over the next years and there is a big community behind it and even Symfony Core Team members maintaining the [Symfony integration](https://github.com/thephpleague/flysystem-bundle).

Still keep in mind the `StorageInterface` of Sulu is not very complex and somebody can provide an own bundle to add support for rekalogika integration via an own bundle if they want: https://github.com/sulu/sulu/blob/2.6/src/Sulu/Bundle/MediaBundle/Media/Storage/StorageInterface.php.

So I see rekalogika not as part of the core, still Sulu is Symfony and you can use whatever you want if you need to create custom entities.

**Comment by RafaelKr** [comment]
Thanks for your extensive feedback and pointing me to the right places of Sulu/Flysystem. I completely understand your point, every external dependency can hurt a project in multiple ways.

> So I see rekalogika not as part of the core, still Sulu is Symfony and you can use whatever you want if you need to create custom entities.

TLDR: Their integration currently depends on league/flysystem [^3.16](https://github.com/rekalogika/file-src/blob/0bcd212b80e3bc3876a2a7127a19cf965eef42e9/composer.json#L20C11-L20C26), whereas sulu references version [^1.0](https://github.com/sulu/sulu/blob/beeb08794280cbeee61c3562ed56999cac31d0bf/composer.json#L120C30-L120C35), also on the 3.0 branch. We can run `composer require rekalogika/file-bundle` and it installs  `league/flysystem (3.29.1)`.
Due to the existence of this issue I'm now asking myself, if we will run into any conflicts when using it, meaning we will have Flysystem v3 installed?


Still, while I mentioned the library in my previous comment it's not really about the library (speak: external dependency) itself, more that they implemented many very interesting concepts and abstractions improving the DX a lot when working with files which I didn't saw before in the PHP ecosystem. Don't get me wrong, my comments are just meant to be some inspiration and brain food, I don't want to criticise Sulu in any way.

Professionally working as a developer since 10+ years and 5+ years in the Symfony ecosystem I just wanted to share that I never had the feeling that files are handled in such a nice way throughout the full lifecycle ([Upload](https://rekalogika.dev/file-bundle/file-upload-filepond), [CRUD/File-Management](https://rekalogika.dev/file/file), [Transformation](https://rekalogika.dev/file/derivation), [Download](https://rekalogika.dev/file-bundle/serving-files), ...) before using their file abstractions.
As that library was released last year and didn't get a lot of traction (yet, besides from that it's still very production ready and well-thought out and structured!) I'm trying to bring a little attention to it. I thought those concepts may be also interesting to other parts of the PHP/Symfony ecosystem, in this case Sulu, in the future.

Besides the points mentioned in my previous comment also their file associations are implemented with a very interesting concept. By using [file pointers](https://rekalogika.dev/file/concepts) (filesystem + [key](https://rekalogika.dev/file/concepts#keys-vs-paths), also see [File Association Internal Details](https://rekalogika.dev/file-bundle/advanced/entity-association-internal)) instead of paths, they are completely independent from the underlying storage layer and can be [lazy loaded](https://rekalogika.dev/file/proxy). This allows a project to start out with a local disk and migrate to an object storage later, if more space is required.

Anyways, enough of "praising" that library/their concepts.
You're right, in theory we could add this dependency ourselfs as Sulu fully follows the Symfony structure, which we love and why our team now decided to fully explore Sulu and started building a new platform project on it and liking it so far. But as told in the TLDR we're still not sure if we'll run into any conflicts later on using their library.

Edit: Added documentation links to "file lifecycle" examples and file associations section

**Comment by alexander-schranz** [comment]
There are no hard requirement to league flysystem currently so you should not run into conflicts which shouldnt be resolvable on the project side. With Sulu 3.0 we probably will require flysystem 3. So there you probably should also not run into any issues if you implement your own StorageInterface based on the rekalogika classes.

**Comment by mamazu** [comment]
This will be in 3.0 See merge request above.
